### ISSUE TITLE
Content addressability

### ISSUE THREAD

**Issue by mntn-xyz** [PigeonProtocolConsortium/pigeon-spec#4]
Hi, found your project thanks to ploum. Solderpunk (creator of Gemini) has been thinking about something similar using Git as a basis, and one potential issue he identified with the scheme is hyperlinks. Hyperlinks are critical for discovering and sharing content, but in a serverless, network-less, offline-first scheme then linking schemes are challenging. I came up with a first attempt at a solution for this here: https://mntn.xyz/posts/2021-10-25-a-proposal-for-a-git-link-url/

In short you use URNs instead of URLs, with a local resolution service to help you locate the content (or queue it for later fetching). I think this approach could be adapted for Pigeon as well. URNs could look like the following: pigeon:DB_MULTIHASH/OBJECT_MULTIHASH. The hashes alone are probably enough with SHA256, but one could incorporate signatures for extra security if hash collisions are a concern, and for future-proofing.

**Comment by RickCarlino** [comment]
Hey @mntn-xyz I'm glad that Ploum is spreading the word about the project. Like I mentioned in the post to Ploum, the project has been in hibernation as I await for more collaborators to show up, so I really appreciate that you took the time to discuss these ideas.

I do agree that hyperlinks are important and the protocol is in a state where large changes like this could still be made. I was thinking something might be more along the lines of `pigeon:multihash` since objects are content addressed and therefore global in scope. Eg: `pigeon:FILE.FV0FJ0YZADY7C5JTTFYPKDBHTZJ5JVVP5TCKP0605WWXYJG4VMRG` or `pigeon:USER.GGP2VX0ZN41EYXMN81YB0Q4AEKRCVZ5RD1F1PHPY3748HAZSHZC4`. Is this similar to what you were thinking?

**Comment by mntn-xyz** [comment]
I suppose something like a DHT could be used to make it easier to discover some content based on a hash alone; I wonder if a DHT is possible over sneakernet?

My reasoning for including the database multihash up front:

- If you already have a copy of the database in question, but don't have the hashed object, this tells you where to find the object, and also that your copy of the database is likely out of date and you need to acquire an update somehow
- If you don't yet have a copy of the database, it should be easier for people to maintain and pass around an index of known databases (and how to find those databases) than storing and passing a list of every known object multihash

**Comment by mntn-xyz** [comment]
Also, note that this is approaching the problem from the perspective that users may not want to carry copies of every database, but are likely willing to help others locate and request databases based on the hash by passing along an index of databases. This is not specified in the protocol but nothing that I see would prevent it from being added. If this protocol ever became popular, I think it's safe to say that people would eventually run out of space if they took a "greedy" approach and carried every database they came in contact with. They may also not want to carry certain databases if they find the content objectionable, for instance.

**Comment by RickCarlino** [comment]
> I wonder if a DHT is possible over sneakernet?

I think it is. The only difference is that you would need to batch multiple GET/SET operations and avoid confirmation messages (since it could take days to get an OK response).

> pass around an index of known databases (and how to find those databases) than storing and passing a list of every known object multihash

This is an idea I was also contemplating, but never had the time to implement.

 >  the perspective that users may not want to carry copies of every database, but are likely willing to help others locate and request databases based on the hash by passing along an index of databases. 

I suppose this might be a defining characteristic of Pigeon- it assumes that you have easy access to very large harddrives. I have plans to eventually build a "garbage collector" for when the blob count gets too high. There could be different levels of cleanliness:

 * Level 1: Keep only the blobs that my local feed references.
 * Level 2: Keep my blobs, plus blobs referenced in feeds I follow.
 * Level 3: Keep blobs that are owned by me, my friends, and their friends.
 
For an offline-only protocol, I think "blob greed" is a feature rather than a shortcoming. If connectivity is scarce, the protocol needs to be prepared with as many blobs as possible. I have seen the benefits of this in action with my SSB client- I often grumble that my `.ssb` dir consumes multiple GBs. But when I found myself without internet, it was very convenient to explore profiles and content of freinds-of-friends.

Feel free to ask more quesitons- it's very fun to see renewed interest in physical peer-to-peer.

**Comment by mntn-xyz** [comment]
In light of the "keep it all" approach you take then that makes sense. However, with the implementation of garbage collection, you may need a reference to the original database. Otherwise hyperlinks lose their permanence. (Unless that is desired, if you are building a more ephemeral protocol.)

Hypothetical: you stumble across a document that references an older document (database unspecified). You wouldn't know what database that document referred to. If you've turned on garbage collection, and you're interested in retrieving that old document, you'd need to reload all of the databases you have, unless the database includes some kind of index file that lists known hashes, but I didn't see this in the spec.

I suppose there is a tradeoff here between relative permanence and long-term hyperlinking on one hand ("library" model), and lightweight ephermerality on the other hand ("newsstand" model). Both are valid mental models. The former brings with it some added complexity and heavier storage requirements; the growing pool of information is useful for deep research/reference, but is also possibly overwhelming both for storage needs and for people who just want to keep up with recent events and conversations.

**Comment by RickCarlino** [comment]
@mntn-xyz 

 > with the implementation of garbage collection, you may need a reference to the original database.

Here's my idea: in Pigeon, the database is a series of key/value pairs that reference:

 * Blobs
 * Messages
 * Other users/databases
 * inline strings

When garbage collection runs, the local node will iterate over all messages in every feed it cares about.
As it iterates, it would collect all multihashes into a large set. It would then delete all multihashes that were not contained in the set.

In the case of hyperlinks inside of a blob, you would need to extract the links and index them in the database entry for the page.

A pseudocode example:

```
author USER.XYE14NZ0XV2PGG4CZHSZAH8473YPHP1F1DR5ZVCRKEA4Q0BY18NM
depth 123
kind gemini_capsule_page
prev TEXT.E90DY6RABDQ2CJPVQHYQDYH6N7Q46SZKQ0AQ76J6D684HYBRKE4G

document_address:FILE.FV0FJ0YZADY7C5JTTFYPKDBHTZJ5JVVP5TCKP0605WWXYJG4VMRG
link0:FILE.25FYEL7KXIF2ASL3LNHWDEH26A2QAAT5WH562LL4DDVIEWDEH26I
link1:FILE.BKJFBSHYWUWCEOCP2H7GDEH24AFNKBLXDPRAZMXSHTRN7WDEH254
link2:FILE.IYW26QOKWGYR5S2LAC7WDEH254JZU3SFMXJQAKU7HICO4WDEH26E

signature JSPJJQJRVBVGV52K2058AR2KFQCWSZ8M8W6Q6PB93R2T3SJ031AYX1X74KCW06HHVQ9Y6NDATGE6NH3W59QY35M58YDQC5WEA1ASW08
```

In the example above, the links are permanently referenced in the database so they will never be garbage collected. Conversely, if you have specific pages you want to delete, you could just manually delete the blob from the filesystem since the protocol is tolerant of missing data.

SSB already follows this pattern. The example below is an SSB post that contains an embedded image (blob):

```json
{
  "previous": "%guZNKEQdRv8icSXtm1xLBG2RVxk5lcsXNdtiTPPU8QY=.sha256",
  "sequence": 1361,
  "author": "@z2M8msI2EUubNHnrEJncglDIy2/SUd+36jCyJnfeiHk=.ed25519",
  "timestamp": 1610309055624,
  "hash": "sha256",
  "content": {
    "type": "post",
    "text": "#aroundme \n\n![A very long icy path down a hill in a wooded area on an overcast winter day.](&lepxcLZ2llzeYhrhMM0qYMiQnrQXrfnOpmwsGXDwox4=.sha256)\n",
    "mentions": [
      {
        "link": "#aroundme"
      },
      {
        "link": "&lepxcLZ2llzeYhrhMM0qYMiQnrQXrfnOpmwsGXDwox4=.sha256",
        "name": "A very long icy path down a hill in a wooded area on an overcast winter day.",
        "type": "image/png",
        "size": 1049256
      }
    ]
  },
  "signature": "uBUnqNSdigvCTOWVT0UiDmfgyFKG63GYs3M/myF3VT1oOiZxtKCneP6bZpIOJFBSV6S3PGDNxwMHSrCx9BxoBw==.sig.ed25519"
}
```

I am unsure if SSB supports garbage collection, but it certainly has all the information it needs to do it.

 > Hypothetical: you stumble across a document that references an older document (database unspecified). You wouldn't know what database that document referred to.

Like SSB, Pigeon is content-addressed, so knowing who the author of a document is won't actually matter. You either have access to a peer that has the content, or you don't. You can see this reality in SSB. SSB messages multihashes (similar to `TEXT.*` entries in Pigeon) do not contain any reference to the original author. In the example above, the multihash is `%fiXoyZRh1+ocEXzVgsOqi3rz8kcFJND91BlCX5fyEWc=.sha256`, but that does not tell you that the author is `@z2M8msI2EUubNHnrEJncglDIy2/SUd+36jCyJnfeiHk=.ed25519`. The client is still able to retrieve the data (asynchronously) by communicating with their peers.

 > If you've turned on garbage collection, and you're interested in retrieving that old document, you'd need to reload all of the databases you have, unless the database includes some kind of index file that lists known hashes, but I didn't see this in the spec.

It would be the responsibility of the application (rather than the protocol) to make sure that it emits messages that are correctly indexed (eg: inform the protocol of all data required). With that being said, you're right, the protocol is still unclear about a lot of aspects of synchronization. I think this is the last hard problem left to address.

 > tradeoff ... between relative permanence and long-term hyperlinking on one hand ("library" model), and lightweight ephemerality on the other hand ("newsstand" model).

I like this analogy. Pigeon is very much a "library" rather than a newsstand. It can be seen in the immutability of database records and the design tradeoffs made with regard to disk space.

 > possibly overwhelming both for storage needs and for people who just want to keep up with recent events and conversations.

So, storage is definitely one area where the protocol is lenient. When I was considering design tradeoffs, conservation of disk space was the aspect of least concern. With that being said, it is still possible to delete blobs without consequence, since the protocol is always operating in an environment where data travels slowly and a full picture of the data develops asynchronously.

For `USER.*` resources (databases), you could unfollow the user and re-index your local DB, then run a GC sweep.

You can't, however, delete `TEXT.*` entries, but:

 * They are typically only a few bytes long.
 * A newer version of the protocol might be able to circumvent this. It appears that some blockchain projects have devised some clever tricks to minimize the size of a chain of hashed messages without losing the ability to securely verify them. See: "ZK Snarks" and the multitude of projects advertising themselves as "lightweight blockchain" (Mina protocol).
 * In the early days, I was going to try to implement something [like this](https://github.com/AljoschaMeyer/bamboo) to enable partial verification of messages without downloading the entire chain, but I shelved the idea in the name of time. I can revisit this idea in a v2 protocol.

I hope this answers your questions. Feel free to ask more if you have any. I very much enjoy seeing outside interest in the project after a year away from it.

**Comment by mntn-xyz** [comment]
Your suggestion for hyperlinking makes sense and would solve the issue by telling clients that certain resources should be "pinned" due to active references. When root nodes are manually deleted, or when old root nodes age out and haven't been referenced by newer objects, the tree could be pruned at that point.

In some hypothetical future where there is just way too much Pigeon data out there, I can see clients offering an interface for pruning the tree manually. I agree that this should be managed by the user/client and not the protocol.

> With that being said, you're right, the protocol is still unclear about a lot of aspects of synchronization. I think this is the last hard problem left to address.

Yes, I think that this is the last big missing piece. I'll think about it and see if I can contribute any ideas. I've been following a number of decentralization projects over the last few years, but almost all of them are overly complex and would not be a good fit for low powered/salvaged devices. I was interested by this project because it reflects some of the permacomputing aesthetics that I appreciate from projects like Gemini and Collapse OS. The protocol is simple enough to be handled by a microcontroller, and the lack of encryption means that messages can legally be transmitted over amateur radio bands. There's a lot of potential here!

Thanks for the link to bamboo, that's an interesting project and could be the basis for partial sync as you mentioned.

Unless you'd prefer to close it, I'll leave this issue open until links find their way in to the protocol. I'll see if I can contribute to the documentation when I have time for it; I saw that you were looking for BNF grammar, for instance. (By the way, I tried signing up for the mailing list but it is no longer active.)

**Comment by RickCarlino** [comment]
> In some hypothetical future where there is just way too much Pigeon data out there, I can see clients offering an interface for pruning the tree manually.

Yes, that's always an option. I am hopeful that there might already be a working solution out there, since it seems like some blockchain projects have already solved this problem and it is an area of intense research these days. Although Pigeon is (obviously) not a blockchain, it does bear some similarities and the project can benefit from advancements in that field.

 > this is the last big missing piece. I'll think about it and see if I can contribute any ideas.

I look forward to hearing it.

One idea I have had for a bundle specification:

 * Everything is stored in a ZIP file. Ideally, a version of ZIP that is still supported by older ZIP packages such as PKZip (Like you, I am interested in supporting older machines). I also like Zip because you _can_ encrypt it, but you don't need to.
 * All feeds are packed in as `n.txt`, where `n` is an arbitrary integer from 0...99999999 (to support FAT16/8.3 naming on older machines). This would contain your feed, as well as the feeds of all peers you follow.
 * All blobs are packed in as `n.bin`, with the same numbering rules as above.
 * At the time of export, the database adds a "manifest entry" to its feed. I will explain this idea in its own section.
 * Since everything is content addressed in Pigeon, and because it is a secure protocol, the recipient must figure out the content addresses of all *.bin files on their own. We cannot tell our peers what the content address is, they need to figure that out for themselves (to ensure a friend-of-a-friend is not lying). If I remember correctly, I did not fix this part in the current version of the spec as it exists today. I still need to go back and update the bundle spec. I did not realize this mistake until I built the Go version.

A big unanswered question is if bundles are unicast or multicast. That is to say, do bundles have intended recipients? Having an intended recipient would mean that you could omit previously sent blobs, etc...

Now, here's what I think we could put in a bundle manifest:

 * A list of blobs we have (since we may be holding onto blobs of peers that our peer does not follow).
 * A list of blobs we want (the client could keep a list of cache misses that occurred since the last "sync")
 * A "last thing I heard" report. Eg: Bob has a cached copy of Alice's feed up to message 10. If we also only have message 10, there is no news to report. Conversely, if we have 12 messages from Alice, it would be possible to include a "slice" of Alice's feed instead of the whole feed (give our peer the only 2 messages they are missing from the whole feed).
 * After ingesting a manifest, a well-behaved client will delete the manifest blob since it contains sensitive meta-data. This makes the blob permanently unreachable. Another idea is to create a new `META.*` type that receives special treatment as an "out of band private message".

 > almost all of them ... would not be a good fit for low-powered/salvaged devices.

I am glad you had this in mind also. In my mind, I have been targeting a Windows 3.1/FAT16 environment, possibly with a few modifications, such as an SD-to-hard drive adapter and a healthy amount of logical partitions for blob storage.

 > I was interested .. because it reflects .. permacomputing aesthetics that I appreciate from projects like Gemini and Collapse OS.

I really do like both of those projects, and am quite a Forth enthusiast as well. Once the protocol is more stable, I would like to create a Forth-based implementation, possibly for low-end devices that call out to an external block storage addon for dealing with the large size of the blob library.

 > The protocol is simple enough to be handled by a microcontroller
 
Supposedly there are [ED25519 libs for Arduino](https://rweather.github.io/arduinolibs/classEd25519.html
). MicroSD breakout cards are also very cheap and easy to use (another reason I like to adhere to FAT16 / 8.3 naming conventions).

 > the lack of encryption means that messages can legally be transmitted over amateur radio bands. There's a lot of potential here!

I hadn't considered that! Originally, I just didn't want to deal with that level of planning, but that's a good point to consider.

 > I saw that you were looking for BNF grammar, for instance. (By the way, I tried signing up for the mailing list but it is no longer active.)

When writing a new implementation of Pigeon (I've written two now), I find that the parser has been the most time-consuming part (assuming the target has good Base32/ED25519 support). Having a BNF grammar that can be easily fed into a parser generator would be a huge help. Unfortunately, I could never figure this part out.
